Convert <model-viewer> article from WebFu. (#3289)
GoogleChrome/web.dev: src/site/content/en/blog/model-viewer/index.md

> ---

Reviewer: @jpmedley & @rkochman are you expecting this to be the authoritative post on model-viewer for the foreseeable future?

I'm asking because you're currently using the authoritative URL https://web.dev/model-viewer

Another option would be https://web.dev/introducing-model-viewer if you wanted to save https://web.dev/model-viewer for something else in the future

Author: This is pulling from another site and it works.

title: Introducing the &lt;model-viewer&gt; web component
subhead: Adding 3D models to a website can be tricky; &lt;model-viewer&gt; is as easy as writing HTML.
authors:
  - joemedley
date: 2019-02-06
updated: 2020-06-25
hero: hero.jpg
alt: A 3D model of a space suit.
description: |
  Adding 3D models to a website is tricky for a variety of reasons including
  the hosting issues and the high bar of 3D programming. That's why we're
  introducing the &lt;model-viewer&gt; web component to let you use 3D models
  declaratively.
tags:
  - blog # blog is a required tag for the article to show up in the blog.
  - 3d
  - model-viewer
---

Adding 3D models to a website is tricky. 3D models ideally will be shown in
a viewer that can work responsively on all browsers including smartphones,
desktop, or even the new head-mounted displays. The viewer should support
progressive enhancement for performance and rendering quality. It should support
use cases on all devices ranging from older, lower-powered smartphones to newer

> devices that support augmented reality. It should stay up to date with current

Reviewer: ```suggestion
devices that support augmented reality. It should stay up-to-date with current
```

Author: This is a prepositional phrase, not an adjective. Therefore, no hyphens.

technologies. It should be performant and accessible. However, building such a
viewer requires specialty 3D programming skills, and can be a challenge for web
developers that want to host their own models instead of using a third-party
hosting service.

> <style>

Reviewer 2: If possible, I'd like to avoid adding inline styles to articles. It's possible that over time as we change the site these styles will break.

  #mv-demo {
    background-color: #36b6fe;
    max-height: 400px;
    height: 400px;
    width: 100%;
  }
</style>

To help with that, the `<model-viewer>` web component, which just released
[version 1.0](https://modelviewer.dev/), lets you declaratively add a 3D model
to a web page, while hosting the model on your own site. The web component
supports responsive design and use cases like augmented reality on some devices,
and it includes features for accessibility, rendering quality, and

> interactivity.  The goal of the component is to enabble adding 3D models to your

Reviewer: ```suggestion
interactivity.  The goal of the component is to enable adding 3D models to your
```

website without understanding the underlying technology and platforms.

{% Aside %}
We're always [updating and
improving](https://github.com/google/model-viewer/releases)
`<model-viewer>`. Check out the [`<model-viewer>`
homepage](https://modelviewer.dev/) to explore what &lt;model-viewer&gt; 1.0 can do.
{% endAside %}

## What is a web component?

A web component is a custom HTML element built from standard web platform
features. A web component behaves for all intents and purposes like a standard
element. It has a unique tag, it can have properties and methods, and it can
fire and respond to events. In short, you don't need to know anything special to
use any web component, much less `<model-viewer>`.

In this article, I will show you things that are particular to `<model-viewer>`.
If you're interested in learning more about web components
[webcomponents.org](https://www.webcomponents.org/) is a good place to start.

## What can &lt;model-viewer> do?

I'll show you a few current capabilities of `<model-viewer>`. You'll get a great

> experience today, and `<model-viewer>` will get better over time as Google adds

Reviewer 3: ```suggestion
experience today, and `<model-viewer>` will get better over time as the team adds
```
We welcome contributions from the community, so better to not say Google here.

new features and improves rendering quality. The examples I'm provided are just
to give you a sense of what it does. If you want to try them there are

> installation and usage instructions on

Author: ```suggestion
installation and usage instructions [on
```

> installation and usage instructions in [its GitHub
> repo](https://modelviewer.dev/).

Reviewer 3: ```suggestion
installation and usage instructions on
modelviewer.dev](https://modelviewer.dev/).
```

### Basic 3D models

Embedding a 3D model is as simple as the markup below. By
using glb files, we've ensured that this component will work on any major
browser.

```html
<!-- Import the component -->
<script type="module" src="https://unpkg.com/@google/model-viewer/dist/model-viewer.min.js"></script>
<script nomodule src="https://unpkg.com/@google/model-viewer/dist/model-viewer-legacy.js"></script>
<!-- Use it like any other HTML element -->
<model-viewer id="mv-demo" shadow-intensity="1" src="./spacesuit.glb"
alt="A 3D model of an astronaut" auto-rotate camera-controls
poster="./spacesuit.jpg"></model-viewer>
```

That code renders like this:

<script type="module" src="https://unpkg.com/@google/model-viewer/dist/model-viewer.min.js"></script>
<script nomodule src="https://unpkg.com/@google/model-viewer/dist/model-viewer-legacy.js"></script>
<model-viewer id="mv-demo" shadow-intensity="1" src="https://modelviewer.dev/shared-assets/models/Astronaut.glb" alt="A 3D model of an astronaut" auto-rotate camera-controls poster="./spacesuit.jpg" ></model-viewer>

> ### Adding motion and interactivity

Reviewer: I said in the email that web.dev can't support embedding third-party script directly into a page

Author: >The `<model-viewer>` web component, which just released [version 1.0](https://modelviewer.dev/), lets you declaratively add a 3D model to a web page.

Then I don't know what to do here. The component is designed to be used exactly as I have done here. We're announcing the component that can be used on any page while simultaneously saying we don't trust it on our page.

Reviewer: Embedding it as a Glitch is the only way you'll get this interactive demo on web.dev

I got the ball rolling on a Glitch: https://glitch.com/~model-viewer-example

You'll want to tweak the CSS so that the demo fills the viewport

Reviewer: We trust model-viewer, we don't trust unpkg

Hosting model-viewer on web.dev also isn't an option because if we're going to ask for features from the engineering team, I would rather them focus on features that will benefit all of our content (such as getting the table of contents landed) rather than a one-off thing like this

Reviewer 2: If you're worried about glitch startup times, another option would be to add a model-viewer sample to https://github.com/GoogleChrome/samples. Those samples get hosted on GitHub pages.

So this dialog example https://github.com/GoogleChrome/samples/tree/gh-pages/dialog
Is available at https://googlechrome.github.io/samples/dialog/index.html
Which you could iframe as:

```
<iframe src="https://googlechrome.github.io/samples/dialog/index.html">
```

similar to how we iframe Glitch. This would probably give you a speedier uptime.

Reviewer: @jpmedley it sounds like iframe-ing the demonstration is another way forward. Do you want to go with that?

The `auto-rotate` and `camera-controls` attributes provide motion and user
control. Those aren't the only possible attributes. See the documentation for [a

> complete list of](https://modelviewer.dev/).

Reviewer 3: ```suggestion
complete list of attributes](https://modelviewer.dev/).
```

```html
<model-viewer src="assets/Astronaut.glb" controls auto-rotate
poster="./spacesuit.jpg">
```

To show your users that it's a 3D model, and not just an image, you can provide
some preload animation by using script to switch between multiple posters.

```html/3-7
<model-viewer id="toggle-poster" src="assets/Astronaut.glb" controls
auto-rotate poster="assets/poster2.png"></model-viewer>
<script>
    const posters = ['poster.png', 'poster2.png'];
    let i = 0;
    setInterval(() =>
        $('#toggle-poster').setAttribute('poster', `assets/${posters[i++ %
2]}`), 2000);
</script>
```

### Responsive Design

The component handles some types of responsive design, scaling for both mobile
and desktop. It can also manage multiple instances on a page and uses
[Intersection
Observer](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver)
to conserve battery power and GPU cycles when a model isn't visible.

<figure class="w-figure">
  <img src="./responsive-spacesuit.png" alt="Multiple spacesuit images representing responsiveness.">
  <figcaption class="w-figcaption">Multiple spacesuit images representing responsiveness.</figcaption>
</figure>

## Looking Forward

> [Install `<model-viewer>` and give it a try](https://modelviewer.dev/) The

Reviewer 3: ```suggestion
[Install `<model-viewer>` and give it a try](https://modelviewer.dev/). The
```

project team wants `<model-viewer>` to be useful to you, and wants your input on

> its future. That's not to say they don't have ideas. So give it a try and let us

Reviewer 3: ```suggestion
its future. So give it a try and let us
```
I would either remove this sentence or link to the roadmap (like the original).
